tests/integration/security/citadel/namespace_targeting_test.go

```diff
@@ -0,0 +1,86 @@
+package citadel
+
+import (
+	"fmt"
+	"istio.io/istio/pkg/test"
+	"istio.io/istio/pkg/test/framework/components/environment/kube"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	v1 "k8s.io/api/core/v1"
+	mv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/citadel"
+)
+
+const (
+	testNsPrefix       = "targeting-test"
+	serviceAccountName = "sample-service-account"
+)
+
+// TestNamespaceTargeting verifies that Citadel respects the designated namespace targeting labels.
+// current implementation is closely tied with Kubernetes concepts, but this is difficult to avoid
+// until Citadel itself is less closely tied with Kubernetes concepts
+func TestNamespaceTargeting(t *testing.T) {
+	framework.NewTest(t).Run(func(ctx framework.TestContext) {
+
+		citadel.NewOrFail(t, ctx, citadel.Config{Istio: ist})
+		env := ctx.Environment().(*kube.Environment)
+
+		// Label combinations
+		type citadelLabels map[string]string
+		testCases := map[string]struct {
+			labels   citadelLabels
+			targeted bool
+		}{
+			"override label set to false, env label matches => should not generate secret": {
+				labels: citadelLabels{
+					"ca.istio.io/override": "false",
+				},
+				targeted: false,
+			},
+			"no override label present, env label matches => should generate secret": {
+				labels: citadelLabels{
+					"ca.istio.io/env": "istio-system", // how can we get this from config?
+				},
+				targeted: true,
+			},
+			"no override label present, no env label present, default is active => should generate secret": {
+				labels:   citadelLabels{},
+				targeted: true,
+			},
+		}
+
+		for name, tc := range testCases {
+			ctx.NewSubTest(name).
+				Run(func(ctx framework.TestContext) {
+					doTest(t, env, ctx, tc.labels, tc.targeted)
+				})
+		}
+	})
+}
+
+func doTest(
+	t test.Failer, env *kube.Environment, ctx framework.TestContext, nsLabels map[string]string, targeted bool) {
+	ns := namespace.NewOrFail(t, ctx, testNsPrefix, false, nsLabels)
+
+	// create service account in newly created namespace
+	serviceAccount := &v1.ServiceAccount{
+		ObjectMeta: mv1.ObjectMeta{
+			Name: serviceAccountName,
+		},
+	}
+	env.GetServiceAccount(ns.Name()).Create(serviceAccount)
+	time.Sleep(time.Second * 5)
```

Comment:
It is better to use a smart check instead of always waiting for 5 seconds. For example, we can add a retry function to check service account before timeout. Or we can add some metrics into Citadel and check the metric values.

Reply:
+1, the hard thing here is that we are testing for something not happening. So when targeted == true, we should just poll until the secret exists, and fail if it doesn't exist within 5s. But for targeted == false, not sure what else we can do?

Reply:
This is very similiar to the accessor component in `kube.go`, search for `WaitUntilPodsAreReady`.